File: level3/ex14.js
function solution(operations) {
    const MAX = 1;
    const MIN = 0;
    
    function append(q, num) {
        q.push(parseInt(num));
        q.sort((a, b) => a - b);
        return q;
    }
    
    function remove(type, q) {
        if (type === MAX) q.pop();
        else if (type === MIN) q.shift();
        return q;
    }
    
    function get(type, q) {
        if (type === MAX) return q[q.length-1];
        else if (type === MIN) return q[0];
    }
    
    let q = [];
    
    for (const operation of operations) {
        const [op, num] = operation.split(' ');
        if (op === 'I') q = append(q, num);
        else if (num === '1') q = remove(MAX, q);
        else if (num === '-1') q = remove(MIN, q);
    }
    
    if (q.length === 0) return [0, 0];
    else return [get(MAX, q), get(MIN, q)];
}

File: level3/ex14.py
from heapq import heappush, heappop, nlargest, heapify

def solution(operations):
    MAX = 1
    MIN = 0
    
    def remove(type, heap):
        if len(heap) == 0:
            return heap
        if type == MAX:
            heap = nlargest(len(heap), heap)[1:]
            heapify(heap)
        elif type == MIN:
            heappop(heap)
        return heap
    
    def get(type, heap):
        if type == MAX: 
            return nlargest(len(heap), heap)[0]
        elif type == MIN: 
            return heappop(heap)
    
    heap = []
    
    for operation in operations:
        op, num = operation.split(' ')
        if op == 'I': 
            heappush(heap, int(num))
        elif num == '1': 
            heap = remove(MAX, heap)
        elif num == '-1': 
            heap = remove(MIN, heap)
                
    if len(heap) == 0: 
        return [0, 0]
    else: 
        return [get(MAX, heap), get(MIN, heap)]